core/src/saros/concurrent/management/HeartbeatDispatcher.java

```diff
@@ -1,72 +1,106 @@
 package saros.concurrent.management;
 
+import java.util.Map.Entry;
 import java.util.concurrent.ScheduledThreadPoolExecutor;
 import java.util.concurrent.TimeUnit;
+import org.apache.log4j.Logger;
 import saros.activities.JupiterActivity;
+import saros.activities.SPath;
+import saros.concurrent.jupiter.Algorithm;
+import saros.concurrent.jupiter.internal.Jupiter;
 import saros.concurrent.jupiter.internal.text.NoOperation;
 import saros.repackaged.picocontainer.Startable;
 import saros.session.AbstractActivityProducer;
 import saros.session.ISarosSession;
+import saros.session.User;
 import saros.synchronize.UISynchronizer;
 import saros.util.NamedThreadFactory;
 
 /**
- * This class sends periodically JupiterOperations to ack other clients Activities.
+ * This class generates and sends periodically Jupiter-NoOperations to acknowledge operations.
  *
- * <p>This fixes a logical memory leak, that is present in the Jupiter-implementation, if a client
- * does not participate in the session actively (usually Saros/S). This causes other clients to hold
- * JupiterActivities indefinitely to be able to merge them at all times. Sending 'NoOperation's
- * fixes this issue by acking received activities though the operation's vectorTime.
+ * <p>See
+ * https://www.researchgate.net/publication/220876978_High-Latency_Low-Bandwidth_Windowing_in_the_Jupiter_Collaboration_System
+ *
+ * <p>Some fine points: In our system, messages must be saved until they are acknowledged by the
+ * other party, since they may be needed in order to fix up incoming messages. Normally, these
+ * acknowledgments are piggy-backed on traffic going the other way. However, it is possible for the
+ * traffic to a window to be one-sided (e.g., for a status display window being periodically
+ * updated). Therefore, each side must periodically generate explicit acknowledgments (i.e. no-op
+ * messages) to prevent the outgoing queues from growing forever.
  */
 public class HeartbeatDispatcher extends AbstractActivityProducer implements Startable {
-  private ScheduledThreadPoolExecutor heartbeatScheduledExecutor;
+
+  private static final Logger log = Logger.getLogger(HeartbeatDispatcher.class);
+
   private final ISarosSession session;
   private final UISynchronizer uiSynchronizer;
-  private final JupiterClient jupiter;
+  private final JupiterClient jupiterClient;
+
+  private ScheduledThreadPoolExecutor heartbeatScheduledExecutor;
```

Comment:
Renaming and moving variables and similar changes do make this a lot harder to review then necessary because functional differences are harder to spot.

Reply:
+1 In general, it would be nice to split off refactorings into separate commits.